### ISSUE TITLE
Accuracy is not 1 as expected in very simple case with single categorical variable

### ISSUE THREAD

**Issue by ismael-elatifi** [catboost/catboost#602]
catboost version: 0.12.0
Operating System: Windows

The issue is described and demonstrated by the code below.

```Python
from catboost import CatBoostClassifier, Pool

def fit_print_accuracy(model, col_cat, target):
    model.fit(col_cat, target)#col_cat is only a single categorical column
    print(model.eval_metrics(Pool(col_cat, label=target, cat_features=[0]), metrics=["Accuracy"]))
    
col_cat = ['A','A','B','B','C','C','D','D']
target =  [ 1,  1,  0,  0,  1,  1,  0,  0]

# each category has always the same target value
# so 100% accuracy can in theory be reached with only one tree with only the following optimal split : 
# categories 'B' and 'D' goes to the left and categories 'A' and 'C' goes to the right

model = CatBoostClassifier(num_trees=1, depth=1, cat_features=[0])

fit_print_accuracy(model, col_cat, target) # Accuracy should be 1 but is 0.5

# duplicate last row
col_cat.append('D')
target.append(0)
fit_print_accuracy(model, col_cat, target) # Accuracy is 1 as expected

# duplicate last row
col_cat.append('D')
target.append(0)
fit_print_accuracy(model, col_cat, target) # Accuracy should still be 1 but is now 0.8

# duplicate last row
col_cat.append('D')
target.append(0)
fit_print_accuracy(model, col_cat, target) # Accuracy is 1 as expected
```

**Comment by Goorman** [comment]
The reason is boosting_type, which is 'Ordered' by default for small datasets. Ordered boosting computes target statistics for categorical feature splits based only on certain specific subset of objects, thus reducing overfitting (read catboost paper for details). 

If you set boosting_type in model constructor to 'Plain', catboost will build trees the same 'plain' way as other libraries do, and will achieve 1.0 accuracy score.

**Comment by ismael-elatifi** [comment]
Thanks for the reply @Goorman . Indeed accuracy is one with boosting_type="Plain".
But I still found the results with boosting_type="Ordered" a bit strange because the accuracy is not decreasing from 1 to 0.8 just by duplicating the last row. I would expect the accuracy to stay at 100%.

**Comment by annaveronika** [comment]
I'm looking through the old issues. Sorry for not answering! The answer is in explanation of how categorical features handling works. We generate new features based on categorical ones, which are representing estimates of probability of having positive value given that the object is of this category.

But we are not using this particular object when we are calculating the feature value for this object. We are using only the objects before this one in some random permutation. This helps a lot with fighting overfitting, because you are not memorizing the data.

Here is a paper that explains this approach in more details:
http://learningsys.org/nips17/assets/papers/paper_11.pdf
You need Part 2 "Categorical features".

Hope it helps!

And thanks for your issue!
